daphne/src/roles/helper.rs

```diff
@@ -216,42 +217,104 @@ pub trait DapHelper<S>: DapAggregator<S> {
-            loop {
-                if retry_count < 1 {
-                    // we need to prevent an attacker from keeping this loop running for too long,
-                    // potentialy enabling an DOS attack.
-                    return Err(DapAbort::BadRequest(
-                        "AggregationJobContinueReq contained too many replays".into(),
-                    ));
+        const RETRY_COUNT: u32 = 3;
+        let mut input_transitions = agg_job_cont_req.transitions;
+        let mut output_transitions: Vec<(ReportId, Option<TransitionVar>)> = input_transitions
+            .iter()
+            .map(|t| &t.report_id)
+            .cloned()
+            .map(|id| (id, None))
+            .collect::<Vec<_>>();
+        for _ in 0..RETRY_COUNT {
+            let (agg_share_span, agg_job_resp) = task_config.vdaf.handle_agg_job_cont_req(
+                task_id,
+                task_config,
+                &state,
+                agg_job_cont_req.round,
+                &agg_job_id,
+                &input_transitions,
+            )?;
+
+            let put_shares_result = self
+                .try_put_agg_share_span(task_id, task_config, agg_share_span)
+                .await;
+
+            for t in agg_job_resp.transitions {
+                for (id, var) in &mut output_transitions {
+                    if *id == t.report_id {
+                        *var = Some(t.var);
+                        break;
+                    }
                 }
-                retry_count -= 1;
-                let (agg_share_span, agg_job_resp) = task_config.vdaf.handle_agg_job_cont_req(
-                    task_id,
-                    task_config,
-                    &state,
-                    |id| replayed_reports.contains(id),
-                    &agg_job_id,
-                    &agg_job_cont_req,
-                    &metrics,
-                )?;
-
-                let out_shares_count = agg_share_span.report_count().try_into().unwrap();
-                let replayed = self
-                    .try_put_agg_share_span(task_id, task_config, agg_share_span)
-                    .await?;
-
-                if let Some(replayed) = replayed {
-                    replayed_reports.extend(replayed);
-                    metrics.agg_job_cont_restarted_inc();
-                } else {
-                    break (out_shares_count, agg_job_resp);
+            }
+
+            let inc_restart_metric = Once::new();
+            for (_, result) in put_shares_result {
+                match result {
+                    // we got 0 replays
+                    (Ok(replays), reports) if replays.is_empty() => {
+                        // remove successfull reports in this batch as they are finished
+                        input_transitions
+                            .retain(|t| reports.iter().all(|(id, _)| *id != t.report_id));
+                        // set those reports as finished in the output_transitions
```

Comment:
I'd like code comments to consistently use proper punctuation.
```suggestion
                        // Set those reports as finished in the output_transitions.
```